src/components/SafeVersionProvider/index.jsx

```diff
@@ -0,0 +1,133 @@
+// @flow
+import * as React from 'react'
+import { useSelector } from 'react-redux'
+import semverLessThan from 'semver/functions/lt'
+import satisfies from 'semver/functions/satisfies'
+import semverValid from 'semver/functions/valid'
+
+import Modal from '~/components/Modal'
+import { getGnosisSafeInstanceAt } from '~/logic/contracts/safeContracts'
+import { getCurrentMasterContractLastVersion } from '~/logic/safe/utils/safeVersion'
+import UpdateSafeModal from '~/routes/safe/components/Settings/UpdateSafeModal'
+import { safeParamAddressFromStateSelector } from '~/routes/safe/store/selectors'
+
+type TSafeVersion = {
+  currentVersion: string,
+  featuresEnabled: string[],
+  lastVersion: string,
+  needsUpdate: boolean,
+  upgradeSafe: () => void,
+}
+
+export const SafeVersionContext = React.createContext<TSafeVersion>({
+  currentVersion: '',
+  featuresEnabled: [],
+  lastVersion: '',
+  needsUpdate: false,
+  upgradeSafe: () => {},
+})
+
+const SafeVersionProvider = ({ children }: { children: React.Node }) => {
```

Comment:
Could you please explain why we need this provider and why version info can't be put into redux store?

Reply:
We have many places where we need to use version info: components, actions, middlewares and getting info from context in all those places may be tricky imo

Reply:
My reasoning was to have one single responsible for the whole app version-related stuff.

A Provider felt like the natural option to me. Having the option to even trigger an upgrade from any place in the app, the provider contains the upgrade Modal.

And I'm being explicit/conscious when I need to consume any version-related value.

Reply:
I feel that the right place would be safe model inside the redux store, what do you think?

Reply:
any update on this?

Reply:
oh, didn't see this comment.

Having `currentVersion`, `latestVersion`, `needsUpdate` and `featuresEnabled` in the store can be an option, but what about the `upgradeSafe` callback?

I mean, with the provider we have all in one place, it's simple to use with hooks (a bit verbose with classed components tho, but our intention is to move all to function components).

And whenever you switch safes, the provider will update and notify the required consumers.

Honestly, I'm not seeing the advantage of redux over a provider here.

Reply:
My main point is that soon we'll have 3 sources for safe versions. We have the provider, we have functions in `utils/safeVersion.js` and I am about to add it to the redux store in my offchain signatures PR because I need safe version info in an action. This is not good. Also keep in mind, we're using redux, it's meant to be a single source of state for our application.

About `upgradeSafe` callback - it could be an action

Reply:
how I imagine the flow to be:
1. Whenever safe is loaded to the interface, we can check it's version and save it inside safe record
2. Fetch it whenever user goes to safe page
3. On app load, we fetch latest master contract version and save into redux something like `state.safes.latestMasterContractVersion`
4. Any consumer can fetch this information/trigger upgrade

Reply:
@mikheevm changes in place: https://github.com/gnosis/safe-react/pull/570/commits/af0661d8f5207b5db00093de83af7562ba2e1f4c, https://github.com/gnosis/safe-react/pull/570/commits/4a820ae1c6868ace56d986a38ada8d6ea0cf6f3a, https://github.com/gnosis/safe-react/pull/570/commits/c03c7b8e9f79bd776a2be72a6fe721b916bb1361, https://github.com/gnosis/safe-react/pull/570/commits/b4f6c65155f1eec2d2ea7faa34c24d74cf25cd88, and https://github.com/gnosis/safe-react/pull/570/commits/173f42810fbc178c7e8d37ca1ac5fa642684eb4e